### PR TITLE
Refactor action traits

### LINKED ISSUE #1967: Simplify component implementations
Remove the following default hook implementations from component bases

- `decorate_action`
- `schedule_thread`
- `get_required_capabilities`
- `is_target_valid`

and replace those with an adaptive traits implementation.

### PR DESCRIPTION
**Pull request by hkaiser**
Refactor traits implementations to simplify component code …  

- refactored traits: is_target_valid, decorate_action, and schedule_thread
- this relies on branch migrate_component 

This fixes #1967

### CODE REVIEW

**Comment by hkaiser** [comment]
Essentially, please review the last commit only, everything else is being reviewed here: #1966

File: examples/quickstart/data_actions.cpp [hunk 1]
```diff
@@ -22,24 +22,6 @@ struct plain_data
     }
 
     static bool is_target_valid(hpx::naming::id_type const& id) { return true; }
```
Comment (sithhell) [inline examples/quickstart/data_actions.cpp]: This can be removed as well, i guess.

Comment (sithhell) [inline examples/quickstart/data_actions.cpp]: My bad, it is a special implementation...

**Comment by sithhell** [comment]
LGTM!
Sidenote:
Having docs that explain the defaults and what those functions mean would be nice.  But I guess that would need proper component documentation to exist in the first place ;)

**Comment by hkaiser** [comment]
I'll add some comments.

**Pull request merged**